pinot-plugins/pinot-batch-ingestion/pinot-batch-ingestion-spark-3/src/main/java/org/apache/pinot/plugin/ingestion/batch/spark3/SparkSegmentMetadataPushJobRunner.java

```diff
@@ -130,4 +196,78 @@ public void call(String segmentTarPath)
       });
     }
   }
+
+  private List<String> getSegmentsToPush(PinotFS outputDirFS, URI outputDirURI) {
+    String[] files = listFiles(outputDirFS, outputDirURI);
+    return Arrays.stream(files).filter(file -> file.endsWith(Constants.TAR_GZ_FILE_EXT)).collect(Collectors.toList());
+  }
+
+  private void setupFileSystems() {
+    List<PinotFSSpec> pinotFSSpecs = _spec.getPinotFSSpecs();
+    for (PinotFSSpec pinotFSSpec : pinotFSSpecs) {
+      PinotFSFactory.register(pinotFSSpec.getScheme(), pinotFSSpec.getClassName(), new PinotConfiguration(pinotFSSpec));
+    }
+  }
+
+  private URI validateOutputDirURI(String outputDir) {
+    URI outputDirURI;
+    try {
+      outputDirURI = new URI(outputDir);
+      if (outputDirURI.getScheme() == null) {
+        outputDirURI = new File(outputDir).toURI();
+      }
+    } catch (URISyntaxException e) {
+      throw new RuntimeException("outputDirURI is not valid - '" + outputDir + "'");
+    }
+    return outputDirURI;
+  }
+
+  private String[] listFiles(PinotFS outputDirFS, URI outputDirURI)
+      throws RuntimeException {
+    try {
+      return outputDirFS.listFiles(outputDirURI, true);
+    } catch (IOException e) {
+      throw new RuntimeException("Unable to list all files under outputDirURI - '" + outputDirURI + "'");
+    }
+  }
+
+  private void setupTableConfigURI() {
+    if (_spec.getTableSpec().getTableConfigURI() == null) {
+      if (_spec.getPinotClusterSpecs() == null || _spec.getPinotClusterSpecs().length == 0) {
+        throw new RuntimeException("Missing property 'tableConfigURI' in 'tableSpec'");
+      }
+      PinotClusterSpec pinotClusterSpec = _spec.getPinotClusterSpecs()[0];
+      String tableConfigURI = SegmentGenerationUtils.generateTableConfigURI(pinotClusterSpec.getControllerURI(),
+          _spec.getTableSpec().getTableName());
+      _spec.getTableSpec().setTableConfigURI(tableConfigURI);
+    }
+  }
+
+  private void executePostUpload(Map<URI, String> uriToLineageEntryIdMap) {
+    try {
+      ConsistentDataPushUtils.postUpload(_spec, uriToLineageEntryIdMap);
+    } catch (Exception e) {
+      ConsistentDataPushUtils.handleUploadException(_spec, uriToLineageEntryIdMap, e);
```

Comment:
Not related to this PR -
if upload fails, then we have to repeat the entire ingestion since revert will remove the segments?
https://github.com/apache/pinot/blob/c6c0fa243b7be407c37749e45512f156297e271a/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java#L195

Would it make sense to have some retry mechanism to workaround transient failures without having to repeat everything?

Reply:
postUpload() calls endReplaceSegments() which has built in retry for handle-able errors. https://github.com/apache/pinot/blob/c6c0fa243b7be407c37749e45512f156297e271a/pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/ConsistentDataPushUtils.java#L168

So it's probably safe to assume by now (when we get to this code) that we want to just clean up / revert instead of trying again